### PR TITLE
ZK-4761: MVVM provide option to disable viewmodel annotation caches

### PR DESCRIPTION
**Pull request by DevChu**

### CODE REVIEW

File: zkbind/src/org/zkoss/bind/impl/BinderImpl.java [hunk 1]
```diff
@@ -2122,11 +2130,15 @@ protected void doExecute(Component comp, String command, Map<String, Object> com
 	private Method getCommandMethod(Class<?> clz, String command, CommandMethodInfoProvider cmdInfo,
 			Map<Class<?>, Map<String, CachedItem<Method>>> cache, int commandParamCount, boolean isGlobal) {
 		Map<String, CachedItem<Method>> methods;
-		synchronized (cache) {
-			methods = cache.get(clz);
-			if (methods == null) {
-				methods = new HashMap<String, CachedItem<Method>>();
```
Comment (b952426) [inline zkbind/src/org/zkoss/bind/impl/BinderImpl.java:2128]: fake empty map

File: zcommon/src/org/zkoss/util/EmptyCacheMap.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * An empty cache map.
+ *
+ * @author jameschu
+ */
+public class EmptyCacheMap<K, V> extends AbstractMap<K, V> implements Cache<K, V> {
+	private static final long serialVersionUID = 20210218L;
```
Comment (scribetw) [inline zcommon/src/org/zkoss/util/EmptyCacheMap.java:26]: yymmddhhmmss

File: zcommon/src/org/zkoss/util/EmptyCacheMap.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	@Override
+	public void setMaxSize(int maxsize) {
+		// do nothing
+	}
+
+	@Override
+	public Set<K> keySet() {
+		return Collections.EMPTY_SET;
```
Comment (scribetw) [inline zcommon/src/org/zkoss/util/EmptyCacheMap.java:55]: Coolections.emptySet()

File: zkbind/src/org/zkoss/bind/impl/BinderImpl.java [hunk 4]
```diff
@@ -147,16 +148,23 @@
...
+	private static final String DISABLE_METHOD_CACHE_PROP = "org.zkoss.bind.disableMethodCache";
+
+	static {
+		DISABLE_METHOD_CACHE = Boolean.parseBoolean(Library.getProperty(DISABLE_METHOD_CACHE_PROP, "false"));
+	}
+
+	private static final Map<Class<?>, List<Method>> _initMethodCache = DISABLE_METHOD_CACHE ? new EmptyCacheMap() : new CacheMap<Class<?>, List<Method>>(600,
 			CacheMap.DEFAULT_LIFETIME); //class,list<init method>
-	private static final Map<Class<?>, List<Method>> _destroyMethodCache = new CacheMap<Class<?>, List<Method>>(600,
+	private static final Map<Class<?>, List<Method>> _destroyMethodCache = DISABLE_METHOD_CACHE ? new EmptyCacheMap() : new CacheMap<Class<?>, List<Method>>(600,
```
Comment (scribetw) [inline zkbind/src/org/zkoss/bind/impl/BinderImpl.java:161]: Can be a singleton and use ConcurrentMap

File: zktest/test/java/org/zkoss/zktest/zats/test2/F96_ZK_4761Test.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+		} catch (Exception e) {
+			fail("check cache error: " + e.getMessage());
+		}
+	}
+
+	private void checkMap(Object obj, String fieldName) throws Exception {
+		assertEquals(0, ((Map) getField(obj, fieldName)).size());
+	}
+
+	private <T> T getField(Object obj, String fieldName) throws Exception {
```
Comment (scribetw) [inline zktest/test/java/org/zkoss/zktest/zats/test2/F96_ZK_4761Test.java:54]: move reflection to VM

**Review by scribetw — CHANGES_REQUESTED** [review]

File: zcommon/src/org/zkoss/util/EmptyCacheMap.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import java.util.AbstractMap;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * An empty cache map.
+ *
+ * @author jameschu
```
Comment (scribetw) [inline zcommon/src/org/zkoss/util/EmptyCacheMap.java:23]: since

**Review by scribetw — APPROVED** [review]

**Review by plijyfes — APPROVED** [review]

**Pull request merged**